### PR TITLE
New dedicated instance resource of APIG service

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
- For the APIG dedicated service, we need to support resource of instances, applications, APIs, environments, etc.
- As a basic container, instance can manage the following resources:
  - API lifecycle.
  - Environment variables.
  - Request throttling.
  - Monitoring and alarm.
  - Access control.
  - VPC channels.
  - Signature keys.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
reference: #1198 

**Special notes for your reviewer**:

**Release note**:

```release-note
1. New dedicated instance release, the following edition are supporte:
  - BASIC, PROFESSIONAL, ENTERPRISE, PLATINUM.
2. Add three acc test for instance:
  - basic configuration and update(name, description, security group and maintenance time).
  - egress binding, unbinding and updation.
  - ingress binding, unbinding and updation.
3. Document supported.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed
Basic test
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccApigInstanceV2_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccApigInstanceV2_basic -timeout 360m -parallel 4
=== RUN   TestAccApigInstanceV2_basic
=== PAUSE TestAccApigInstanceV2_basic
=== CONT  TestAccApigInstanceV2_basic
--- PASS: TestAccApigInstanceV2_basic (456.28s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       456.334s
```
Egress test
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccApigInstanceV2_egress'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccApigInstanceV2_egress -timeout 360m -parallel 4
=== RUN   TestAccApigInstanceV2_egress
=== PAUSE TestAccApigInstanceV2_egress
=== CONT  TestAccApigInstanceV2_egress
--- PASS: TestAccApigInstanceV2_egress (560.10s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       560.145s
```
Ingress test
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccApigInstanceV2_ingress'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccApigInstanceV2_ingress -timeout 360m -parallel 4
=== RUN   TestAccApigInstanceV2_ingress
=== PAUSE TestAccApigInstanceV2_ingress
=== CONT  TestAccApigInstanceV2_ingress
--- PASS: TestAccApigInstanceV2_ingress (567.09s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       567.177s
```

### CODE REVIEW

File: docs/resources/apig_instance.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+* `eip_id` - (Optional, String) Specifies the eip ID associated with the APIG dedicated instance.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - ID of the APIG dedicated instance.
+* `create_time` - Time when the APIG instance is created, in RFC-3339 format.
+* `status` - Status of the APIG dedicated instance.
+* `supported` - The supported features of the APIG dedicated instance.
```
Comment (ShiChangkuo) [inline docs/resources/apig_instance.md]: the attribute is defined as `supported_features` in the schema actually.

File: huaweicloud/resource_huaweicloud_apig_instance.go [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+	}
+	d.SetId(v.Id)
+	err = waitForApigInstanceCreateCompleted(d, client)
+	if err != nil {
+		return fmt.Errorf("Error waiting for APIG v2 dedicated instance (%s) to become running: %s", d.Id(), err)
+	}
+	return resourceApigInstanceV2Read(d, meta)
+}
+
+func setApigAvailableZoneIds(d *schema.ResourceData, resp instances.Instance) error {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_apig_instance.go]: please add more comments to explain why we should do this.

File: docs/resources/apig_instance.md (hunk 1 above)
Comment (Lance52259) [reply docs/resources/apig_instance.md]: fixed

File: huaweicloud/resource_huaweicloud_apig_instance.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_apig_instance.go]: the function description has been added.

File: huaweicloud/resource_huaweicloud_apig_instance.go [hunk 3]
```diff
@@ -154,8 +154,14 @@ func buildMaintainEndTime(maintainStart string) (string, error) {
 	return fmt.Sprintf("%02d:00:00", (num+4)%24), nil
 }
 
+// Method buildRFC3339Timestamp is used to unify the time format to RFC-3339 and return a time string.
+func buildRFC3339Timestamp(timestamp int64) string {
+	createTime := time.Unix(timestamp, 0)
+	return createTime.Format(time.RFC3339)
+}
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_apig_instance.go]: move to utils directory ? @Jason-Zhang9309 What do you think about this ?

File: docs/resources/apig_instance.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  Changing this will create a new APIG dedicated instance resource.
+
+* `subnet_id` - (Required, String, ForceNew) Specifies an ID of the VPC subnet used to create the APIG dedicated
+  instance.
+  Changing this will create a new APIG dedicated instance resource.
+
+* `security_group_id` - (Required, String) Specifies an ID of the security group to which the APIG dedicated instance
+  belongs to.
+
+* `available_zone_ids` - (Optional, List, ForceNew) Specifies an array of available zone IDs for the APIG dedicated
```
Comment (ShiChangkuo) [inline docs/resources/apig_instance.md]: suggest to rename `available_zones`

File: docs/resources/apig_instance.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  Changing this will create a new APIG dedicated instance resource.
+
+* `description` - (Optional, String) Specifies the description about the APIG dedicated instance.
+  The description contain a maximum of 255 characters and the angle brackets (< and >) are not allowed.
+
+* `enterprise_project_id` - (Optional, String, ForceNew) Specifies an enterprise project ID.
+  This parameter is required for enterprise users.
+  Changing this will create a new APIG dedicated instance resource.
+
+* `maintain_time` - (Optional, String) Specifies a start time of the maintenance time window in the format 'xx:00:00'.
```
Comment (ShiChangkuo) [inline docs/resources/apig_instance.md]: suggest to rename `maintain_begin` and add `maintain_end` as an attribute.

File: huaweicloud/resource_huaweicloud_apig_instance.go (hunk 3 above)
Comment (ShiChangkuo) [reply huaweicloud/resource_huaweicloud_apig_instance.go]: move to utils.go will be more clearly.
also we can get a better name like FormatTimeStamp

File: docs/resources/apig_instance.md (hunk 4 above)
Comment (Lance52259) [reply docs/resources/apig_instance.md]: all parameters (in resource, test and document) and functions have been updated.

File: docs/resources/apig_instance.md (hunk 5 above)
Comment (Lance52259) [reply docs/resources/apig_instance.md]: the maintain windows has been updated.

**Pull request merged**